src/merkletree/index.js

```diff
@@ -0,0 +1,134 @@
+const { fork } = require("child_process")
+
+const {
+    BUILD_TREE,
+    SUCCESS,
+    ERROR,
+    deserialiseTree,
+} = require("./worker")
+
+const MerkleTree = require("./api")
+
+function wrapErrorObject(err, message) {
+    let newError
+    if (err instanceof Error) {
+        newError = Object.create(err)
+    } else {
+        newError = new Error()
+    }
+    if (err.stack) {
+        // copy stack
+        newError.stack = err.stack
+    }
+    // prefix with new message
+    newError.message = [message, err.message].filter(Boolean).join("\n")
+    return newError
+}
+
+class Worker {
+    async send(message) {
+        return new Promise((resolve, reject) => {
+            let result
+            if (this.proc) {
+                this.proc.kill()
+            }
+            this.proc = fork(require.resolve("./worker"))
+            this.proc.once("exit", (code) => {
+                if (result) {
+                    resolve(result)
+                    return
+                }
+
+                reject(new Error(`Worker exited with code: ${code}`))
+            })
+            this.proc.once("error", reject)
+            this.proc.once("message", (message) => {
+                result = message
+            })
+            this.proc.send(message)
+        })
+    }
+
+    async sendAction({ type, payload }) {
+        return this.send({ type, payload })
+            .catch((error) => ({
+                type: ERROR,
+                payload: error,
+            }))
+    }
+
+    async buildTree(treeContents) {
+        const result = await this.sendAction({
```

Comment:
is `sendAction` re-entry protection? Like, everyone who calls buildTree later would get the result from same worker? `send` does generate a new promise every time, but I guess it's a promise that resolves when worker exits?

Reply:
> is sendAction re-entry protection?

Not really, mainly just a wrapper around `ChildProcess#send` that defines an "Action" as the interface, and attaches the appropriate event listeners and error handling.

> I guess it's a promise that resolves when worker exits?

Correct